### PR TITLE
Add specviz2d example notebook and load miri s2d files

### LINKED ISSUE #532: specviz2d fails to load miri s2d file
Attempting to load a MIRI s2d file into Specviz2d fails with a missing header keyword error.  It looks like the 2d viewer successfully populates but during the extraction of a 1d spectrum, it fails with `KeyError: "Keyword 'WAVEND' not found."`

This file is from pipeline version `0.16.1` and it may be the case that this is outdated, and/or it's not quite a level 3 science product supported by jdaviz.  But one question I have is if jdaviz is expected to work with intermediate JWST data products, i.e. calibration level 2x and below?

I have the download link for the file in question.

To reproduce:
```
file='/Users/bcherinka/Downloads/jw00623-c1019_t015_miri_clear_s2d.fits'
s = Specviz2d()
s.load_data(spectrum_2d=file)
```
Full traceback
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-8-fb42f9f6889b> in <module>
----> 1 s.load_data(spectrum_2d='/Users/bcherinka/Downloads/jw00623-c1019_t015_miri_clear_s2d.fits')

~/Work/git/havok2063/jdaviz/jdaviz/configs/specviz2d/helper.py in load_data(self, spectrum_2d, spectrum_1d, spectrum_1d_label, spectrum_2d_label, show_in_viewer)
    131             # Collapse the 2D spectrum to 1D if no 1D spectrum provided
    132             if spectrum_1d is None:
--> 133                 self.app.load_data(spectrum_2d,
    134                                    parser_reference="spec2d-1d-parser",
    135                                    data_label=spectrum_1d_label,

~/Work/git/havok2063/jdaviz/jdaviz/app.py in load_data(self, file_obj, parser_reference, **kwargs)
    312                 # If the parser returns something other than known, assume it's
    313                 #  a message we want to make the user aware of.
--> 314                 msg = parser(self, file_obj, **kwargs)
    315 
    316                 if msg is not None:

~/Work/git/havok2063/jdaviz/jdaviz/configs/specviz2d/plugins/parsers.py in spec2d_1d_parser(app, data_obj, data_label, show_in_viewer)
     47         except ValueError:
     48             flux = u.Quantity(np.sum(data, header['DISPAXIS']-1))
---> 49         step = (header["WAVEND"] - header["WAVSTART"])/len(flux)
     50         spectral_axis = np.arange(header["WAVSTART"], header["WAVEND"],
     51                                   step) * u.Unit("m")

~/anaconda3/envs/havokve/lib/python3.8/site-packages/astropy/io/fits/header.py in __getitem__(self, key)
    154             keyword = key
    155 
--> 156         card = self._cards[self._cardindex(key)]
    157 
    158         if card.field_specifier is not None and keyword == card.rawkeyword:

~/anaconda3/envs/havokve/lib/python3.8/site-packages/astropy/io/fits/header.py in _cardindex(self, key)
   1714 
   1715         if not indices:
-> 1716             raise KeyError(f"Keyword {keyword!r} not found.")
   1717 
   1718         try:
```

### PR DESCRIPTION
**Pull request by javerbukh**
### Description

As said in the title. Loads data if "WAVEND" and "WAVESTART" header extensions are not present by using pixel values. There is a more accurate solution that @havok2063 mentioned [here](https://github.com/spacetelescope/jdaviz/issues/532#issuecomment-830210418) but that requires using `jwst` as a dependency, which might require some discussion.

Fixes #532 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -48,7 +48,7 @@ Specviz
 
 Other Changes and Additions
 ---------------------------
-
+- MIRI s2d files can now be loaded into Specviz2d.
```
Comment (pllim) [inline CHANGES.rst:51]:
```suggestion
- MIRI s2d files can now be loaded into Specviz2d. [#915]
```

File: jdaviz/configs/specviz2d/plugins/parsers.py [hunk 2]
```diff
@@ -46,9 +46,15 @@ def spec2d_1d_parser(app, data_obj, data_label=None, show_in_viewer=True):
             flux = np.sum(data, header['DISPAXIS']-1)*u.Unit(header["BUNIT"])
         except ValueError:
             flux = u.Quantity(np.sum(data, header['DISPAXIS']-1))
-        step = (header["WAVEND"] - header["WAVSTART"])/len(flux)
-        spectral_axis = np.arange(header["WAVSTART"], header["WAVEND"],
-                                  step) * u.Unit("m")
+
+        if "WAVEND" in header and "WAVSTART" in header:
+            step = (header["WAVEND"] - header["WAVSTART"])/len(flux)
```
Comment (pllim) [inline jdaviz/configs/specviz2d/plugins/parsers.py:51]: So, this assumes linear range. I am not familiar with the specs, is this assumption always true?
```suggestion
            step = (header["WAVEND"] - header["WAVSTART"]) / flux.size
```

File: jdaviz/configs/specviz2d/plugins/parsers.py [hunk 3]
```diff
@@ -46,9 +46,15 @@ def spec2d_1d_parser(app, data_obj, data_label=None, show_in_viewer=True):
...
+            spectral_axis = np.arange(header["WAVSTART"], header["WAVEND"],
+                                      step) * u.Unit("m")
```
Comment (pllim) [inline jdaviz/configs/specviz2d/plugins/parsers.py:53]: Just use the unit instance directly.
```suggestion
                                      step) * u.m
```

File: jdaviz/configs/specviz2d/plugins/parsers.py [hunk 4]
```diff
@@ -46,9 +46,15 @@ def spec2d_1d_parser(app, data_obj, data_label=None, show_in_viewer=True):
...
+        else:
+            # u.Unit("m") is used if WAVEND and WAVSTART are present so
+            # we use it here as well, even though the actual unit is pixels
+            spectral_axis = np.arange(1, len(flux)+1, 1) * u.Unit("m")
```
Comment (pllim) [inline jdaviz/configs/specviz2d/plugins/parsers.py:57]: I am not sure how I feel about casting meter when you know it is in pixels, but I am no domain expert.
```suggestion
            spectral_axis = np.arange(1, flux.size + 1, 1) * u.m
```

File: notebooks/Specviz2dExample.ipynb [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b962e74a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "filename = \"/Users/javerbukh/Documents/jdaviz_dev/data/data_for_mosviz/det_image_seq5_MIRIMAGE_P750Lexp1_s2d.fits\""
```
Comment (pllim) [inline notebooks/Specviz2dExample.ipynb:65]: Make the dirname general.
```suggestion
    "filename = \"/path/to/det_image_seq5_MIRIMAGE_P750Lexp1_s2d.fits\""
```

**Review by pllim — COMMENTED** [review]
General review, mostly nitpicks.

**Comment by pllim** [comment]
p.s. I just added a milestone, so the next time you push, the change log check should pass.

**Comment by pllim** [comment]
p.p.s. You probably also want to update the new example notebook narrative to follow the style that @eteq set in  #870

**Comment by pllim** [comment]
I noticed 3 things while running this new notebook. If out of scope, are there existing tickets to track these already. If not, maybe we should open new tickets.

1. The toolbar is not flush against right side anymore. @kecnry , do you know why? Is this related to your recent work on the UI?
2. I don't know what the region selection is supposed to do. I thought if I select something, it will collapse the selected 2D spectrum region and display the collapsed 1D in the bottom viewer, but the selection does nothing.
3. Is the top viewer X-axis labels hidden on purpose? Personally I would like to see them but maybe PO had the opposite idea before I joined the team.

image

**Comment by kecnry** [comment]
> 1. The toolbar is not flush against right side anymore. @kecnry , do you know why? Is this related to your recent work on the UI?

This is only in the voila app, not notebooks or lab, right?  I have noticed that as well, already created a ticket, and can confirm it is present in main.

**Comment by pllim** [comment]
> This is only in the voila app, not notebooks or lab, right? 

No, I saw this in Lab. I actually cannot run viola on my setup for a completely different reason.

**Comment by javerbukh** [comment]
@pllim Thank you for the comments! In regards to point 2, I'm not sure how it is supposed to work. Ideally it will work the same as in Mosviz, although I remember @rosteen having to change a few things in the case that a user enters a 2d spectra only. In regards to point 3, the X-axis is in-fact hidden on purpose. 

I'll update the notebook now.

**Comment by pllim** [comment]
> Ideally it will work the same as in Mosviz

Are you saying it needs the Collapse plugin like this?

https://github.com/spacetelescope/jdaviz/blob/86fce7bc773e79339de4f5f45a1f5f67ce21ede2/jdaviz/configs/cubeviz/cubeviz.yaml#L20

If that is the case, I don't see it imported for Specviz2d.

**Comment by javerbukh** [comment]
Nope, this is the line of code I was referencing: https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/configs/specviz2d/helper.py#L131-#L136

**Comment by pllim** [comment]
But that logic is not hooked up to event listener for subset selection. Did I miss something here?

**Comment by javerbukh** [comment]
I don't think I can answer that question then. Does subset selection work with a different dataset in specviz2d?

**Comment by pllim** [comment]
> Does subset selection work with a different dataset in specviz2d?

I have no idea. I haven't fired up Specviz2d until today to test this PR. I mean, the selection works, but it does not appear to be doing anything. 🤷

**Review by pllim — APPROVED** [review]
I think my relevant comments are addressed. My remaining comments are probably out of scope, so approving. Thanks!

**Comment by pllim** [comment]
I guess I was naively expecting the selection to generate its own line profile, not unlike this in Ginga:

Screenshot 2021-09-29 173817

**Comment by havok2063** [comment]
Specviz2D was meant as a pared-down Mosviz stopgap solution until a fuller more dynamic Mosviz could be implemented.  So I'm not sure if we ever checked that the various features and plugins worked as expected, or appropriately, in this configuration.  Since it's Mosviz-lite, I would guess the features and plugins would work similarly to how they do in Mosviz.

File: notebooks/Specviz2dExample.ipynb [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ec94837b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "specviz2d.load_data(filename)"
+   ]
+  },
```
Comment (kecnry) [inline notebooks/Specviz2dExample.ipynb:86]: Can we do this programmatically using `astropy.utils.data.download_file` as in other notebooks?  I tried swapping it out, but it seems to not recognize the file format and unlike `specviz.load_spectrum`, `specviz2d.load_data` does not seem to accept the format kwarg - so perhaps this is out of scope?

Comment (pllim) [reply notebooks/Specviz2dExample.ipynb:86]: I think I know why... The link brings you to the *Box page* where the data is, but it is not a direct link to the file itself. If @javerbukh fixed that Box link to be direct link to file, direct download will work.

Comment (pllim) [reply notebooks/Specviz2dExample.ipynb:86]: Also, I accidentally found a bug. I will open a PR to this PR.

File: notebooks/Specviz2dExample.ipynb [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+   "outputs": [
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "374197be0c4a43bbbe2f4987f5a46a75",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "Application(components={'g-viewer-tab': '<template>\\n  <component :is=\"stack.container\">\\n    <g-viewer-tab\\n …"
```
Comment (pllim) [inline notebooks/Specviz2dExample.ipynb:53]: I think you forgot to clear output?

File: notebooks/Specviz2dExample.ipynb [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "b962e74a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#filename = \"/path/to/det_image_seq5_MIRIMAGE_P750Lexp1_s2d.fits\"\n",
```
Comment (pllim) [inline notebooks/Specviz2dExample.ipynb:89]: The associated cells need to be cleaned up. They look messy and inconsistent now.

Comment (javerbukh) [reply notebooks/Specviz2dExample.ipynb]: I always forget I need to manually save the notebooks after I'm done editing them...

**Review by pllim — APPROVED** [review]
LGTM. Thanks!

**Review by rosteen — APPROVED** [review]
